### DISCUSSION TITLE
How to get visible text from flow part,please?

### DISCUSSION THREAD

**Discussion by effapp** [OrchardCMS/OrchardCore#16916] [Q&A]
How to get visible text from flow part,please?

**Comment by Piedone** [comment]
You can start by rendering the content item to HTML:

```c#
// Inject IContentManager
var contentItem = await _contentManager.GetAsync(contentItemId);
// IContentItemDisplayManager and IUpdateModelAccessor (or the current controller instance)
var shape = await _contentItemDisplayManager.BuildDisplayAsync(contentItem, _updateModelAccessor.ModelUpdater);
// IDisplayHelper
var htmlContent = await _displayHelper.ShapeExecuteAsync(shape);
```

**Reply by effapp** [reply]
Hi @Piedone ,
I only need a string visible text of flow part in content item, not all of content item and not html.

**Reply by Piedone** [reply]
You can use placement to hide everything you don't need with a custom display type. Then you'll get the HTML of just Flow Part. That you can process with something like HtmlAgilityPack to get what you need.

**Reply by effapp** [reply]
> You can start by rendering the content item to HTML:
> 
> ```cs
> // Inject IContentManager
> var contentItem = await _contentManager.GetAsync(contentItemId);
> // IContentItemDisplayManager and IUpdateModelAccessor (or the current controller instance)
> var shape = await _contentItemDisplayManager.BuildDisplayAsync(contentItem, _updateModelAccessor.ModelUpdater);
> // IDisplayHelper
> var htmlContent = await _displayHelper.ShapeExecuteAsync(shape);
> ```

Hi @Piedone ,
     var shape = await _contentItemDisplayManager.BuildDisplayAsync(contentItem, _updateModelAccessor.ModelUpdater);

How to get the FlowPart Shape from variant shape,please?

**Reply by Piedone** [reply]
Inspect `shape` in the debugger, and in a property that shape should be there. However, I'd rather recommend what I told above with a custom display type.

**Reply by effapp** [reply]
I don't know how to  use placement to hide everything that don't need with a custom display type yet.

Inspect shape in the debugger, can't tell which one is flow part

**Comment by effapp** [comment]
In Content-Article.cshtml, "@await DisplayAsync(Model.Content.AbstractContent)" can render correctly but in Content-Article.Summary.cshtml "@await DisplayAsync(Model.Content.AbstractContent)" render nothing,  why?

Note: Model.Content.AbstractContent is FlowPart

**Reply by ns8482e** [reply]
You might not be placing anything at location AbstractContent for Summary

**Reply by effapp** [reply]
Model.Content.AbstractContent is null in Summary View but is not null in details view like Content-Article.cshtml

**Reply by ns8482e** [reply]
Use console log to log your shapes between Summary and Detail, your display driver might not be placing any shape in case of Summary view

**Reply by ns8482e** [reply]
If it's null you need to place using placement for Summary view

**Reply by effapp** [reply]
I didn't have display driver, I use default display driver.

**Comment by sebastienros** [comment]
As I moved the issue @effapp created to discussions I am discovering this one here too ... So it seems we still don't have a working solution.

**Reply by effapp** [reply]
Does anybody have a working solution, please?
